CodegenProposal.md

```diff
@@ -0,0 +1,1474 @@
+
+The schema types have a number of functions. 
+
+* Include metadata that allows the `GraphQLExecutor` and runtime type checking on `TypeCase`s to operate. 
+* Can be extended to provide cache key computation for types to configure the normalized cache.
+* Used for interacting with the cache for manual read/write functionality.
+* Used to create mock objects for generated `SelectionSet`s to be used in unit tests.
+
+
+These schema types can be included directly in your application target, or be generated into a separate shared library that can be used across modules in your application.
+
+Schema types are implemented as `class` objects, not `struct`s. They will use reference type semantics and are mutable within a cache transaction.
+
+## `Object` Types
+
+For each concrete type declared in your schema and referenced by any generated operation, an `Object` subclass is generated.
+
+## `Interface` Types
+
+For each interface type declared in your schema and referenced by any generated operation, an `Interface` subclass is generated. Interfaces wrap an underlying `Object` type and ensure that only objects of types that they are only initialized with a wrapped object of a type that implements the interface.
+
+## `Union` Types
+
+For each union type declared in your schema and referenced by any generated operation, a `UnionType` enum is generated. `UnionType` enums have cases representing each possible type in the union. Each case has an associated value of the `Object` type represented by that case. `UnionType` enums are referenced as fields by being wrapped in a `Union<UnionType>` enum that provides access to the underlying `UnionType` and unknown cases. See [Handling Unknown Types](#handling-unknown-types) for more information.
+
+## `Schema` Metadata
+
+A `SchemaConfiguration` object will also be generated for your schema. This object will have a function that maps the `Object` types in your schema to their `__typename` string. This allows the execution to convert data (from a network response from the cache) to the correct `Object` type at runtime.
+
+For an example of generated schema metadata see [AnimalKindgomAPI/Schema.swift](Tests/ApolloCodegenTests/AnimalKingdomAPI/ExpectedGeneratedOutput/Schema.swift).
+
+# `EnumType` Generation
+
+Enums will be generated for each `enum` type in the schema that is used in any of the operations defined in your application. These enums will conform to a simple `EnumType` protocol. When used as the type for a field on a `SelectionSet`, these enums will be wrapped in the generic `GraphQLEnum`.
+
+```swift
+enum RelativeSize: String, EnumType {
+  case LARGE
+  case AVERAGE
+  case SMALL
+}
```

Comment:
probably worth clarifying that unlike in current codegen, we'll respect capitalization from the schema.

Reply:
Will do! Thanks

Reply:
We should confirm that's the case actually. Going to depend on if the `graphql-js` library does that.

Reply:
Confirmed that is does respect case of the schema. I'll update that.

Reply:
Yeah my recollection was there was some weird shit in the TS code that I could never track down that did this, it wasn't GQL-js